presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java

```diff
@@ -7056,4 +7060,60 @@ public TypeAndEstimate(Type type, EstimatedStatsAndCost estimate)
             this.estimate = requireNonNull(estimate, "estimate is null");
         }
     }
+
+    private class GreaterThanZeroProcessedDataSizeCheck
+    {
+        private final DistributedQueryRunner queryRunner;
+        private final String query;
+
+        private ResultWithQueryId<MaterializedResult> queryResult;
+        private long processedInputBytes;
+
+        public GreaterThanZeroProcessedDataSizeCheck(DistributedQueryRunner queryRunner, @Language("SQL") String query)
+        {
+            this.queryRunner = queryRunner;
+            this.query = query;
+            runQuery();
+            getProcessedBytes();
+        }
+
+        private void runQuery()
+        {
+            queryResult = queryRunner.executeWithQueryId(getSession(), query);
+        }
+
+        private void getProcessedBytes()
+        {
+            processedInputBytes = getQueryInfo(queryRunner, queryResult).getQueryStats().getProcessedInputDataSize().toBytes();
+        }
+
+        private long checkProcessedBytes()
+                throws Exception
+        {
+            try {
+                assertThat(processedInputBytes).isGreaterThan(0);
+            }
+            catch (AssertionError error) {
+                // If query statistics haven't been updated at the time we issued getQueryInfo(), we will mistakenly see 0 processedInputBytes.
```

Comment:
can't we just retry calling `getQueryInfo` once again in a while?

Wouldn't simple assertEventually work here?

Reply:
When I've investigated similar problems in the past it seemed that sometimes query stats updates get applied seconds after the query finishes, an sometimes they get lost in transfer. I suspect TCP retransmissions to be the cause, but I never spent enough time to confirm this. 

This is why the query itself also needs to be resubmitted - getting stats repeatedly for the same query won't help if  coordinator already finished collecting stats updates for this particular query.

As for using `assertEventually` - It would be easier to understand, but the race condition in each retry would have the same chances of happening. Thanks to RetryDriver separating `onRetry`, `run` and exponential backoff, we can increase sleep times between the query and asking for it's stats, so the race condition has less of a chance of happening.

Reply:
I have hard time with this one. Retry driver does not rally play well with being uses as exponenetially growing `sleep` between two statements. The fact that we need to run query in `catch` clause. As well as in constructor for 0-seconds sleep, makes it hard to follow.

What about sth like:
```java
       ExponentialSleeper sleeper = new ExponentialSleeper(
                new Duration(0, SECONDS),
                new Duration(5, SECONDS),
                new Duration(1, SECONDS),
                1.5);
        assertEventually(new Duration(30, SECONDS), () -> {
            ResultWithQueryId<MaterializedResult> result = queryRunner.executeWithQueryId(
                    getSession(),
                    "SELECT * FROM test_parquet_timestamp_predicate_pushdown WHERE t = TIMESTAMP '2012-10-31 01:00'");
            sleeper.sleep();
            assertThat(getQueryInfo(queryRunner, result).getQueryStats().getProcessedInputDataSize().toBytes()).isGreaterThan(0);
        });
    }

    private static class ExponentialSleeper
    {
        private Duration nextSleepTime;
        private final Duration maxSleepTime;
        private final Duration minSleepIncrement;
        private final double sleepIncrementFactor;

        private ExponentialSleeper(Duration minSleepTime, Duration maxSleepTime, Duration minSleepIncrement, double sleepIncrementFactor)
        {
            this.nextSleepTime = minSleepTime;
            this.maxSleepTime = maxSleepTime;
            this.minSleepIncrement = minSleepIncrement;
            this.sleepIncrementFactor = sleepIncrementFactor;
        }

        public void sleep()
        {
            try {
                Thread.sleep(nextSleepTime.toMillis());
                long incrementMillis = (long) (nextSleepTime.toMillis() * sleepIncrementFactor - nextSleepTime.toMillis());
                if (incrementMillis < minSleepIncrement.toMillis()) {
                    incrementMillis = minSleepIncrement.toMillis();
                }
                nextSleepTime = new Duration(nextSleepTime.toMillis() + incrementMillis, MILLISECONDS);
                if (nextSleepTime.compareTo(maxSleepTime) > 0) {
                    nextSleepTime = maxSleepTime;
                }
            }
            catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new RuntimeException(e);
            }
        }
    }

```

Reply:
I guess this new class may come useful in other places too (maybe)